[doc] Update doc with steps for custom metric collection (#1791)
operator-framework/operator-sdk: doc/user/metrics/README.md

        ...

        // Add to the below struct any other metrics ports you want to expose.

> 	 servicePorts := []v1.ServicePort{

Reviewer: ```suggestion
	servicePorts := []v1.ServicePort{
```

Reviewer: Small nit in the indentation here.

  - update
...
```
### Custom resource specific metrics

By default operator will expose info metrics based on the number of the current instances of an operator's custom resources in the cluster. It leverages [kube-state-metrics][ksm] as a library to generate those metrics. Metrics initialization lives in the `cmd/manager/main.go` file of the operator in the `serveCRMetrics` function. Its arguments are a custom resource's group, version, and kind to generate the metrics. The metrics are served on `0.0.0.0:8686/metrics` by default. To modify the exposed metrics port number, change the `operatorMetricsPort` variable at the top of the `cmd/manager/main.go` file in the generated operator.

### Custom Metrics using Operator-SDK

> To implement custom metric collection using the `metrics package` provided by SDK, follow the steps outlined below:

Reviewer: ```suggestion
Following the steps to implement custom metric collection using the `metrics package` provided by SDK.
```

> 1. Update the Operator: Update the operator to provide sufficient permisisons so that the required resources for prometheus can be created.

Reviewer: What permission is required that it not added by default?

Author: I have removed this part, as the required permissions are added by default.

2. Custom metric collection: Implement the logic for custom metric variables defined for your operator and register the specific metrics with prometheus. 
3. Expose metrics: Utilize the functions provided by SDK to expose metrics and start the metrics server.

> 

Reviewer: IMHO has no need for the above steps since they will be described bellow

> #### Update the Operator

Reviewer: ```suggestion
#### Add RBCA Permissions
```

> Update the operator and provide sufficient permissions to allow creation of Prometheus resources. The required permissions are:

Reviewer: ```suggestion
Ensure that `role.yaml` file has the following permissions. 
```

1. `Service Monitor` : `get` and `create`
2. `roles` and `rolebindings` : `get`, `create`, `delete`, `update`, `list`

The below is a snippet from roles.yaml:
```yaml
- apiGroups:
  - rbac.authorization.k8s.io
  resources:
  - roles
  - rolebindings
  verbs:
  - get
  - create
  - delete
  - update
  - list
  - watch
- apiGroups:
  - monitoring.coreos.com
  resources:
  - servicemonitors
  verbs:
  - get

>   - create

Reviewer: The following part is already present in the default roles added in the Role.yaml, so should it be really required to be added here? Should not better just add what is required to do in order to make it work?  

```
- apiGroups:
  - monitoring.coreos.com
  resources:
  - servicemonitors
  verbs:
  - get
  - create
```
IMHO: we could remove this part.

```

> #### Custom metric collection

Reviewer: ```suggestion
#### Developing Customized Metrics 
```

> Implement the logic to collect specific custom metrics inside the operator code base. For example, the following is a snippet from `Marketplace` operator which implements custom metric collection:

Reviewer: ```suggestion
Following an example of business logic implementation for customized metrics.  
```

Reviewer: Could you please add an example based on the Memcached sample? 

Also, WDYT to add a NOTE with the link for the code source where has this impl in the `Marketplace` project as the following example?

 ```suggestion
...
**NOTE:** To check a further example see it implemented [here][] in the [Marketplace][marketplace] project  
```

Author: Im not sure on whether the metrics collection is completely done on marketplace. I have provided a CounterVec as an example. Users have directly use this inside their operator. I'm not sure on what is required with specific to Memcached sample. Could you please elaborate more on that?

Reviewer: But you added `following is a snippet from `Marketplace`, so would be great to have the link for this piece of code there and the `Marketplace`. Or, it should not be referenced.  is it make sense?

```Go
// Metric variables and metric list that are used to report error. 
var (
	errorMetric = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "error_count",
			Help: "Describes the monotonic count of the error",
		},
		[]string{errorName, errorValue},
	)
    metricsList = []prometheus.Collector{
		errorMetric,
	}
)
```
Register the metrics created above with Prometheus:

```Go
// registerMetrics registers the metrics with prometheus.
func registerMetrics() error {
	for _, metric := range metricsList {
		err := prometheus.Register(metric)
		if err != nil {
			return err
		}
	}
	return nil
}
```

> #### Expose metrics

Reviewer: ```suggestion
#### Exposing customized metrics
```

The steps provided previously for creating a service and service monitor resource can be followed to expose custom metrics. Additionally, the metrics server has to be manually started instead of providing it as an argument to the controller-runtime manager.
Usage:

> 

Reviewer: Does NOT the following code generated by default in the scalfod project? WDYT about adding an example of a customized metric above using the Memcached sample and them show here just what needs to be done to expose it?

Author: Yes, the code is now generated by default in the scaffold project. But the server has to still be started again. I have provided the code just as an example to do so.

Reviewer: It shows fine.

```Go
import(
        "context"
        "github.com/operator-framework/operator-sdk/pkg/metrics"
)

func main() {

        ...

        // Change the below variables to serve metrics on different host or port.
        const(
            metricsHost = "0.0.0.0"
            metricsPort = 8383
        )

        // Pass metrics address to controller-runtime manager.
        mgr, err := manager.New(cfg, manager.Options{
            Namespace:          namespace,
        })

        ...

        // Add to the below struct any other metrics ports you want to expose.
	    servicePorts := []v1.ServicePort{
		    {Port: metricsPort, Name: metrics.OperatorPortName, Protocol: v1.ProtocolTCP, TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: metricsPort}},
	    }

        // Create Service object to expose the metrics port.
        _, err = metrics.CreateMetricsService(context.TODO(), cfg, servicePorts)
        if err != nil {

>             // TODO (user): handle error

Reviewer: ```suggestion
           log.Error(err, "")

```

        }

        // Start the metrics server

>         http.Handle(MetricsPath, promhttp.Handler())

Reviewer: from where came this MetricsPath? It is not working.

[ownerref-permission]: https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#ownerreferencespermissionenforcement
[ksm]: https://github.com/kubernetes/kube-state-metrics
[controller-metrics]: https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/internal/controller/metrics

### Custom Metrics using Operator-SDK
To implement custom metric collection using the `metrics package` provided by SDK, follow the steps outlined below:
1. Update the Operator: Update the operator to provide sufficient permisisons so that the required resources for prometheus can be created.
2. Custom metric collection: Implement the logic for custom metric variables defined for your operator and register the specific metrics with prometheus. 
4. Expose metrics: Utilize the functions provided by SDK to expose metrics and start the metrics server.

#### Update the Operator

Update the operator and provide sufficient permissions to allow creation of Prometheus resources. The required permissions are:
1. `Service Monitor` : `get` and `create`
2. `roles` and `rolebindings` : `get`, `create`, `delete`, `update`, `list`

The below is a snippet from roles.yaml:

> ```

Reviewer: ```suggestion
```YAML
```

- apiGroups:
  - rbac.authorization.k8s.io
  resources:
  - roles
  - rolebindings
  verbs:
  - get
  - create
  - delete
  - update
  - list
  - watch
- apiGroups:
  - monitoring.coreos.com
  resources:
  - servicemonitors
  verbs:
  - get
  - create

```

#### Custom metric collection

Implement the logic to collect specific custom metrics inside the operator code base. For example, the following is a snippet from `Marketplace` operator which implements custom metric collection:

> ```

Reviewer: ```suggestion
```Go
```

// Metric variables and metric list that are used to report error. 
var (

> 	errorMetric = prometheus.NewCounterVec(

Reviewer: It shows very nice, but IMHO the indentation should equal it is done in the other docs. See here: https://github.com/operator-framework/operator-sdk/blob/master/doc/user/unit-testing.md

Reviewer: Solved

		prometheus.CounterOpts{
			Name: "error_count",
			Help: "Describes the monotonic count of the error",
		},
		[]string{errorName, errorValue},
	)

    metricsList = []prometheus.Collector{
		errorMetric,
	}
)
```
Register the metrics created above with Prometheus:

> ```

Reviewer: ```suggestion
```Go
```

// registerMetrics registers the metrics with prometheus.
func registerMetrics() error {
	for _, metric := range metricsList {
		err := prometheus.Register(metric)
		if err != nil {
			return err
		}
	}
	return nil
}
```
#### Expose metrics
The steps provided previously for creating a service and service monitor resource can be followed to expose custom metrics. Additionally, the metrics server has to be manually started instead of providing it as an argument to the controller-runtime manager.

Usage:

```
import(
        "context"
        "github.com/operator-framework/operator-sdk/pkg/metrics"
)

func main() {

        ...

        // Change the below variables to serve metrics on different host or port.
        const(
            metricsHost = "0.0.0.0"
            metricsPort = 8383
        )

        // Pass metrics address to controller-runtime manager
        mgr, err := manager.New(cfg, manager.Options{
            Namespace:          namespace,
        })

        ...

        // Add to the below struct any other metrics ports you want to expose.

> 	    servicePorts := []v1.ServicePort{

Reviewer: I think it need to be indented.

		    {Port: metricsPort, Name: metrics.OperatorPortName, Protocol: v1.ProtocolTCP, TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: metricsPort}},
	    }

        // Create Service object to expose the metrics port.
        _, err = metrics.CreateMetricsService(context.TODO(), cfg, servicePorts)
        if err != nil {

>             // handle error

Reviewer: ```suggestion
            // TODO (user): handle error
```
